### PR TITLE
Adds radial menu interactions for CIC

### PR DESCRIPTION
**Pull request by breadhunt**
## About The Pull Request
Adds various new interactions on middle-click when using an OW console, if theres any others thatd be useful lmk

Doesn't apply to AI because it's not their niche anymore, can't do overwatch while in a drone + no hands
Also: 
makes OB lase a subtype of CAS to give CIC vision on it, so they can interact with the lase & check its impact zone (it was previously all static)
an OB lased pings anyone in OW
CIC messages now show maptext, like send order does (edited so it's not too annoying w/ the automated ones)

While in OW view, middle-clicking a turf, marine or OB lase will show different options

Turf:
- Orbital spotlight (if possible)
- Message all marines in a ~10 tile radius
- Additional squad actions; messaging all marines in a specific squad, or moving all nearby marines to a specific squad

Marine:
- Message this person
- aSL/demote this person
- Switch this person's squad

OB lase:
- Mark this lase on the minimap for 30 seconds for all marines to view
- Fire OB
## Why It's Good For The Game

CIC has a reputation for being extremely slow and incompetent. Large part of this is mainly because the UI is so clunky.
Radial menus are significantly easier to navigate, especially when there's the option to do it based on location (e.g everyone pushing a certain area, or everyone near the tadpole, or everyone in medbay... etc. etc. etc. etc.)

Cohesion good and improves the quality of the game for both sides
## Changelog
:cl:
add: Radial menus for CIC on middle-click, try it on turfs, marines and OB lases
add: OB being lased pings anyone on CIC to get them to notice
add: CIC messages now show like the Send Order button
balance: OB lase now gives vision to CIC so they can interact with it
/:cl:

### CODE REVIEW

**Comment by DeltaFire15** [comment]
If OB lazes funtion just like a CAS laze does (even being a subtype), why would you ever use a CAS laze instead of just going OB mode and getting free fear factor even if you are just lazing for CAS?

Also maybe this radial menu should be on Ctrl-click instead of MMB if that is compatible, since that would fit well together with the shift-click send order aswell as this being a more "controlled" interaction.

**Comment by XElectricX** [comment]
Legendary PR. Gold star, commander.

**Comment by FosterGrime-Heymynameslouie** [comment]
bless this pr, one of the greatest of all time
speedemrghe

**Comment by Losenis** [comment]
>Quad defiler stacking the separate chems all in one go

image

**Comment by breadhunt** [comment]
> If OB lazes function just like a CAS laze does (even being a subtype), why would you ever use a CAS laze instead of just going OB mode and getting free fear factor even if you are just lazing for CAS?

uhh idk could make it take longer to lase, but afaik CAS/CIC/AI all share vision so

**Comment by Naaanii** [comment]
OB laze always gave slight vision to CIC for them to fire it iirc.

**Comment by Naaanii** [comment]
OB laze can still be fired on, it shouldn't give CIC vision where they shouldn't tbh and CIC must trust the competency of the one lasing. The Good Hits meme should live on.

**Comment by dopamiin0** [comment]
> OB laze can still be fired on, it shouldn't give CIC vision where they shouldn't tbh and CIC must trust the competency of the one lasing. The Good Hits meme should live on.

you overestimate the intelligence of marines. FF OBs will live on.

File: code/datums/jobs/squads.dm [hunk 1]
```diff
@@ -386,30 +386,24 @@
 	var/text = copytext_char(sanitize(message), 1, MAX_MESSAGE_LEN)
 	if(ishuman(sender))
 		var/obj/item/card/id/ID = sender.get_idcard()
-		nametext = "[ID?.rank] [sender.name] transmits: "
+		nametext = "[ID?.rank] [sender.real_name] transmits: "
```
Comment (TiviPlus) [inline code/datums/jobs/squads.dm:389]: Why?

File: code/datums/jobs/squads.dm [hunk 2]
```diff
@@ -386,30 +386,24 @@
...
-	var/text = span_notice("<B>\[Overwatch\]:</b> [format_message(message, sender)]")
-	for(var/i in marines_list)
-		var/mob/living/L = i
-		message_member(L, text, sender)
-
-
-/datum/squad/proc/message_leader(message, mob/living/carbon/human/sender)
-	if(!squad_leader || squad_leader.stat != CONSCIOUS || !squad_leader.client)
-		return FALSE
-	return message_member(squad_leader, span_notice("<B>\[SL Overwatch\]:</b> [format_message(message, sender)]"), sender)
```
Comment (TiviPlus) [inline code/datums/jobs/squads.dm:404]: Is message member used anymore

File: code/game/objects/effects/overlays.dm [hunk 3]
```diff
@@ -227,15 +227,15 @@
...
 	icon_state = "laser_target2"
 
-/obj/effect/overlay/temp/laser_target/OB/Initialize(mapload, effect_duration, named, assigned_squad)
+/obj/effect/overlay/temp/laser_target/cas/OB/Initialize(mapload, effect_duration, named, assigned_squad)
 	. = ..()
 	SEND_GLOBAL_SIGNAL(COMSIG_GLOB_OB_LASER_CREATED, src)
 	GLOB.active_laser_targets += src
 
-/obj/effect/overlay/temp/laser_target/OB/Destroy()
+/obj/effect/overlay/temp/laser_target/cas/OB/Destroy()
```
Comment (TiviPlus) [inline code/game/objects/effects/overlays.dm:238]: This makes cas able to shoot at OB lazes and is treated as a cas laz just copy the camera code I think

File: code/game/objects/machinery/overwatch.dm [hunk 4]
```diff
@@ -50,6 +66,8 @@ GLOBAL_LIST_EMPTY(active_cas_targets)
 	var/datum/action/innate/order/rally_order/send_rally_order
 	///Groundside minimap for overwatch
 	var/datum/action/minimap/marine/external/cic_mini
+	///Ref of the lase that's had an OB warning mark placed on the minimap
+	var/obj/effect/overlay/temp/laser_target/cas/OB/marked_lase
```
Comment (TiviPlus) [inline code/game/objects/machinery/overwatch.dm:70]: Don't see protection against harddel for this

File: code/game/objects/machinery/overwatch.dm [hunk 5]
```diff
@@ -323,41 +341,39 @@ GLOBAL_LIST_EMPTY(active_cas_targets)
...
+				var/input = tgui_input_text(usr, "Please write a message to announce to the squad:", "Squad Message")
 				if(input)
 					current_squad.message_squad(input, usr) //message, adds username
 					if(issilicon(usr))
 						to_chat(usr, span_boldnotice("Message sent to all Marines of squad '[current_squad]'."))
 					visible_message(span_boldnotice("Message sent to all Marines of squad '[current_squad]'."))
 		if("sl_message")
 			if(current_squad && operator == usr)
-				var/input = stripped_input(usr, "Please write a message to announce to the squad leader:", "SL Message")
+				var/input = tgui_input_text(usr, "Please write a message to announce to the squad leader:", "SL Message")
```
Comment (TiviPlus) [inline code/game/objects/machinery/overwatch.dm:358]: Sidenote I noticed this has no IC filter
Todo check also what else stripped input does that tgui doesn't

File: code/game/objects/machinery/overwatch.dm [hunk 6]
```diff
@@ -388,11 +404,33 @@ GLOBAL_LIST_EMPTY(active_cas_targets)
...
-			change_lead()
+			if(!current_squad)
+				to_chat(usr, "[icon2html(src, usr)] [span_warning("No squad selected!")]")
+				return
+			var/sl_candidates = list()
+			for(var/mob/living/carbon/human/H in current_squad.get_all_members())
+				if(istype(H) && H.stat != DEAD && H.mind && !is_banned_from(H.ckey, SQUAD_LEADER))
+					sl_candidates += H
+			var/new_lead = tgui_input_list(usr, "Choose a new Squad Leader", null, sl_candidates)
+			if(!new_lead || new_lead == "Cancel") return
```
Comment (TiviPlus) [inline code/game/objects/machinery/overwatch.dm:415]: Same line ifs are not allowed

File: code/game/objects/machinery/overwatch.dm [hunk 7]
```diff
@@ -559,14 +597,16 @@ GLOBAL_LIST_EMPTY(active_cas_targets)
...
 	if(selected_target)
 		playsound(selected_target.loc,'sound/effects/alert.ogg', 50, 1, 20)  //mostly used to warn xenos as the new ob sounds have a quiet beginning
 
-	addtimer(CALLBACK(src, PROC_REF(send_to_squads), "Transmitting beacon feed..."), 1.5 SECONDS)
-	addtimer(CALLBACK(src, PROC_REF(send_to_squads), "Calibrating trajectory window..."), 3 SECONDS)
 	addtimer(CALLBACK(src, PROC_REF(do_fire_bombard), T, usr), 3.1 SECONDS)
 
+/obj/machinery/computer/camera_advanced/overwatch/proc/alert_lase(datum/source, obj/effect/overlay/temp/laser_target/cas/OB/incoming_laser)
+	to_chat(usr, span_notice("Orbital Bombardment laser detected. Target: [AREACOORD_NO_Z(incoming_laser)]"))
+	usr.playsound_local(usr, 'sound/effects/binoctarget.ogg', 15)
```
Comment (TiviPlus) [inline code/game/objects/machinery/overwatch.dm:608]: Usr usage outside of certain byond related procs is not allowed

File: code/game/objects/machinery/overwatch.dm [hunk 8]
```diff
@@ -559,14 +597,16 @@ GLOBAL_LIST_EMPTY(active_cas_targets)
...
-	send_to_squads("Initializing fire coordinates...")
+	send_to_squads("!!ORBITAL BOMBARDMENT INBOUND AT [get_area(selected_target)]!!")
 	if(selected_target)
 		playsound(selected_target.loc,'sound/effects/alert.ogg', 50, 1, 20)  //mostly used to warn xenos as the new ob sounds have a quiet beginning
 
-	addtimer(CALLBACK(src, PROC_REF(send_to_squads), "Transmitting beacon feed..."), 1.5 SECONDS)
-	addtimer(CALLBACK(src, PROC_REF(send_to_squads), "Calibrating trajectory window..."), 3 SECONDS)
 	addtimer(CALLBACK(src, PROC_REF(do_fire_bombard), T, usr), 3.1 SECONDS)
 
+/obj/machinery/computer/camera_advanced/overwatch/proc/alert_lase(datum/source, obj/effect/overlay/temp/laser_target/cas/OB/incoming_laser)
```
Comment (TiviPlus) [inline code/game/objects/machinery/overwatch.dm:606]: Missing autodoc

File: code/game/objects/machinery/overwatch.dm [hunk 9]
```diff
@@ -559,14 +597,16 @@ GLOBAL_LIST_EMPTY(active_cas_targets)
 		log_attack("[key_name(usr)] fired a [warhead_type]in for squad [current_squad] in [AREACOORD(T)].")
 		message_admins("[ADMIN_TPMONTY(usr)] fired a [warhead_type]for squad [current_squad] in [ADMIN_VERBOSEJMP(T)].")
 	visible_message(span_boldnotice("Orbital bombardment request accepted. Orbital cannons are now calibrating."))
-	send_to_squads("Initializing fire coordinates...")
+	send_to_squads("!!ORBITAL BOMBARDMENT INBOUND AT [get_area(selected_target)]!!")
```
Comment (TiviPlus) [inline code/game/objects/machinery/overwatch.dm:600]: ??

File: code/game/objects/machinery/overwatch.dm [hunk 10]
```diff
@@ -581,43 +621,34 @@ GLOBAL_LIST_EMPTY(active_cas_targets)
 		target.ceiling_debris_check(5)
 		GLOB.marine_main_ship?.orbital_cannon?.fire_ob_cannon(target,user)
 
-/obj/machinery/computer/camera_advanced/overwatch/proc/change_lead()
+/obj/machinery/computer/camera_advanced/overwatch/proc/change_lead(mob/living/carbon/human/H)
```
Comment (TiviPlus) [inline code/game/objects/machinery/overwatch.dm:624]: Better var name would be appreciated

File: code/game/objects/machinery/overwatch.dm [hunk 11]
```diff
@@ -649,56 +680,37 @@ GLOBAL_LIST_EMPTY(active_cas_targets)
...
-		return
 
 	if(!transfer_marine.job)
 		CRASH("[transfer_marine] selected for transfer without a job.")
 
-	if(S != current_squad)
-		return //don't change overwatched squad, idiot.
-
-	if(!istype(transfer_marine) || transfer_marine.stat == DEAD) //gibbed, decapitated, dead
+	if(!istype(transfer_marine)) //gibbed
```
Comment (TiviPlus) [inline code/game/objects/machinery/overwatch.dm:690]: Skipping through removed a lot of checks, 
???

**Review by TiviPlus — CHANGES_REQUESTED** [review]
Good ux change, quick first pass

File: code/datums/jobs/squads.dm (hunk 1 above)
Comment (breadhunt) [reply code/datums/jobs/squads.dm:389]: > Why?

Avoids clutter of '(Unknown) as Urist McOfficer'

File: code/game/objects/effects/overlays.dm (hunk 3 above)
Comment (breadhunt) [reply code/game/objects/effects/overlays.dm:238]: Couldnt get it to work before butI can try again

File: code/game/objects/machinery/overwatch.dm (hunk 9 above)
Comment (breadhunt) [reply code/game/objects/machinery/overwatch.dm:600]: This uses send order text now, avoids clutter & is clearer

Comment (TiviPlus) [reply code/game/objects/machinery/overwatch.dm:600]: I meant why are you changing the wording
the extra !! dont really help for readability so im confused

**Comment by breadhunt** [comment]
Currently finishing this up, can open a new PR if it'd be more convenient

File: code/game/objects/machinery/overwatch.dm (hunk 11 above)
Comment (breadhunt) [reply code/game/objects/machinery/overwatch.dm:690]: The checks were literally just restrictive, I double checked and can't find anything that would go wrong by removing them. As far as I can tell they're only CM relics with 0 actual purpose

File: code/game/objects/machinery/overwatch.dm (hunk 9 above)
Comment (breadhunt) [reply code/game/objects/machinery/overwatch.dm:600]: removed the !!

wording changes make it easier to understand (hopefully)

File: code/datums/jobs/squads.dm (hunk 2 above)
Comment (breadhunt) [reply code/datums/jobs/squads.dm:404]: yea, it's used by messaging functions

**Pull request closed without merging**